apps/openmw/mwsound/soundmanagerimp.cpp

```diff
@@ -53,6 +80,7 @@ namespace MWSound
         , mUnderwaterSound(nullptr)
         , mNearWaterSound(nullptr)
         , mPlaybackPaused(false)
+        , mWorkQueue(new SceneUtil::WorkQueue(1))
```

Comment:
Not sure if it is OK to use different work queues in different places.

Reply:
In general most likely it's not OK, but the problem is when we have different threads working on different stuff OS will schedule them properly. If we start to mix I/O sound related stuff, cell preloading, objects unrefs, navmesh building we would need a much smarter scheduler. Otherwise each type of work might be blocked by another and produce unwanted behaviour like delays. Also mixing CPU bound and I/O bound work in a single thread is not a good idea, because these are different types of resources and it's unwanted to stop navmesh building or unreferencing because the sound file is opened. Yes, a work queue can have multiple threads, but again we'd need much complex logic inside to utilize them properly while we can just use separate objects that are much simpler.

Reply:
Also keep in mind that most of users have not too powerful GPU's - about 2c/4t, which can be already utilized by OpenMW (main thread, draw thread, preloading thread, navmesh updates thread). In this case your new task can be executed on the main thread or on the draw one anyway, so the issue is not really fixed.

Reply:
You benefit from having many more threads than cores as long as you're not triggering contention issues. If most of them are just blocking on I/O access you're not going to have issues or diminishing returns *just* from having additional threads. Doing CPU and I/O bound work on the same thread is much worse than anything you're going to experience from pushing the I/O bound stuff onto a second thread. If there are going to be problems here they're probably going to of the sort that elsid eluded to.

Reply:
That is pretty much how Python's threading works too or any other 'light weight threading' implementation, defer work that is waiting on I/O out to another thread so that the main loop can continue. There is almost no CPU contention here with I/O bound threads.

Reply:
A lot of the time they use state machines or coroutines or green threads instead of OS/hardware threads, but the end result is the same: not blocking on I/O.

Reply:
>  Doing CPU and I/O bound work on the same thread is much worse than anything you're going to experience from pushing the I/O bound stuff onto a second thread.

We do not know how much CPU time ffmpeg itself takes for data processing and how much time it takes just to wait for IO, so it is unclear how it will behave with lack of available CPU cores.

Also keep in mind that WorkQueue with one thread is completely sequential. It means that it can start to execute next work item only after previous one is fully finished. So if there are 5 sounds in queue and it takes 0.1s to handle every file, 5th one can not be handled in next 0.4s, and 0.2s threshold will not actually help.

Reply:
I don't think ffmpeg is doing anything particularly expensive. At the very least it shouldn't be. It's not like it has to do pre-baked compression or resampling or anything.

Yes, the point about how the work queue itself works is valid.